### PR TITLE
Align DASC numeric validation with scalar API

### PR DESCRIPTION
**Pull request by kaix-nv**
## Summary

Addresses both Claude findings on #2396 while retaining CodeRabbit’s oversized-integer fix:

- match the annotated scalar-float API by accepting `numbers.Real` inputs, including NumPy real scalars
- reject all tensor objects before scalar conversion, so multi-element tensors cannot leak RuntimeError
- keep OverflowError normalization for oversized integer inputs
- remove the tensor-scalar compatibility claim and restore the original scalar conversion
- cover multi-element tensors through both exported public entry points

## Validation

- focused DASC suite: 46 passed, 1 optional Megatron skip
- pre-commit hooks on all three changed files
- signed commit with DCO sign-off

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved validation for decay-analysis inputs by requiring real-number values for epsilon and static gate parameters.
  - Invalid non-numeric and tensor-based inputs are now rejected consistently, including zero-dimensional tensor values.
  - Improved numerical consistency when calculating decay horizons by preserving CPU double-precision behavior.
  - Added clearer safeguards against invalid or unsupported input types before analysis calculations are performed.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def compute_gdn_decay_horizons(
     dt_bias_cpu = dt_bias.detach().to(device="cpu", dtype=torch.float64)
 
     decay = -torch.exp(a_log_cpu) * F.softplus(dt_bias_cpu + static_gate_input)
-    horizons = torch.log(torch.as_tensor(epsilon, device="cpu", dtype=torch.float64)) / decay
+    horizons = torch.log(torch.tensor(epsilon, dtype=torch.float64)) / decay
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Fixed in #2398. The horizon numerator now uses math.log(epsilon), so it allocates no tensor and cannot inherit an ambient CUDA/default device. A meta default-device regression verifies the result remains CPU.

[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Fixed in #2398 with the suggested device-free math.log(epsilon) calculation. Added a regression using a non-CPU ambient torch.device context to assert that the documented CPU output contract is preserved.

File: modelopt/torch/sparsity/state_sparsity/config.py [hunk 2]
```diff
@@ -36,19 +37,23 @@
...
         epsilon_is_valid = False
     if not epsilon_is_valid:
         raise ValueError("epsilon must be finite and in (0, 1)")
 
     try:
-        static_gate_input_is_valid = math.isfinite(static_gate_input)
+        static_gate_input_is_valid = isinstance(static_gate_input, Real) and math.isfinite(
+            static_gate_input
+        )
     except (TypeError, ValueError, OverflowError):
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/config.py]: Both suggestions are fixed in #2398. bool is explicitly excluded for epsilon and static_gate_input, tests cover both public paths, and the handlers now catch only the reachable OverflowError case.
